test_runner/performance/test_physical_replication.py

```diff
@@ -0,0 +1,232 @@
+    neon_get_connection_uri,
+    neon_suspend_endpoint,
+    neon_wait_for_operation_to_finish,
+)
+
+
+# Granularity of ~0.5 sec
+def measure_replication_lag(master, replica):
+    start = time.time()
+    master.execute("SELECT pg_current_wal_flush_lsn()")
+    master_lsn = Lsn(master.fetchall()[0][0])
+    while True:
+        replica.execute("select pg_last_wal_replay_lsn()")
+        replica_lsn = replica.fetchall()[0][0]
+        if replica_lsn:
+            if Lsn(replica_lsn) >= master_lsn:
+                return time.time() - start
+        time.sleep(0.5)
+
+
+@pytest.mark.remote_cluster
+@pytest.mark.timeout(0)
+def test_ro_replica_lag(
+    pg_bin: PgBin,
+    neon_api_key: str,
+    neon_api_base_url: str,
+    pg_version: PgVersion,
+):
+    test_duration_min = 5
+    sync_interval_min = 1
+
+    project = neon_create_project(neon_api_key, neon_api_base_url, pg_version)
+    project_id = project["project"]["id"]
+    neon_wait_for_operation_to_finish(neon_api_key, neon_api_base_url, project_id)
+    should_delete = True
+    try:
+        branch_id = project["branch"]["id"]
+        master_connstr = project["connection_uris"][0]["connection_uri"]
+
+        replica = neon_create_endpoint(
+            neon_api_key,
+            neon_api_base_url,
+            project_id,
+            branch_id,
+            endpoint_type="read_only",
+        )
+        neon_wait_for_operation_to_finish(neon_api_key, neon_api_base_url, project_id)
+
+        replica_connstr = neon_get_connection_uri(
+            neon_api_key,
+            neon_api_base_url,
+            project_id,
+            endpoint_id=replica["endpoint"]["id"],
+        )["uri"]
+
+        pg_bin.run_capture(["pgbench", "-i", "-s100", master_connstr])
+        conn_master = psycopg2.connect(master_connstr)
+        cur_master = conn_master.cursor()
+        conn_replica = psycopg2.connect(replica_connstr)
+        cur_replica = conn_replica.cursor()
+
+        master_workload = pg_bin.run_nonblocking(
+            ["pgbench", "-c10", "-T1000", "-Mprepared", master_connstr]
+        )
+        try:
+            replica_workload = pg_bin.run_nonblocking(
+                ["pgbench", "-c10", "-T1000", "-S", replica_connstr]
+            )
+            try:
+                start = time.time()
+                while time.time() - start < test_duration_min * 60:
+                    time.sleep(sync_interval_min * 60)
+                    lag = measure_replication_lag(cur_master, cur_replica)
+                    log.info(f"Replica lagged behind master by {lag} seconds")
+            finally:
+                replica_workload.terminate()
+        finally:
+            master_workload.terminate()
+    except Exception:
+        should_delete = False
+    finally:
+        if should_delete:
+            neon_delete_project(neon_api_key, neon_api_base_url, project_id)
+
+
+@pytest.mark.remote_cluster
+@pytest.mark.timeout(0)
+def test_replication_start_stop(
```

Comment:
It seems that the test does a lot of things, let's add a comment to it

Reply:
Added comment